packages/trail-fastify-plugin/lib/index.js

```diff
@@ -0,0 +1,116 @@
+const fp = require('fastify-plugin')
+const { TrailsManager } = require('@nearform/trail-core')
+const { get } = require('lodash')
+const Ajv = require('ajv')
+
+const { errorsMessages } = require('./schemas/errors')
+
+const environment = get(process, 'env.NODE_ENV', 'development')
+
+function getCustomErrorMessage (schema, path, message) {
+  // Convert a schema path like '#/properties/who/anyOf' to a ref addressing
+  // the custom error type, e.g. 'properties.who.meta.errorType'
+  const ref = path
+    .substring(path.indexOf('/') + 1, path.lastIndexOf('/'))
+    .replace(/\//g, '.')
+    .concat('.meta.errorType')
+  const errorType = get(schema, ref)
+  return errorType ? errorsMessages[errorType] : message
+}
+
+function formatReasons (error, schema) {
+  const { validation } = error
+  return validation.reduce((reasons, item) => {
+    const {
+      dataPath,
+      schemaPath,
+      keyword,
+      message,
+      params
+    } = item
+    const name = dataPath.slice(1)
+    switch (keyword) {
+      case 'required':
+        if (!schemaPath.includes('anyOf')) {
+          reasons[params.missingProperty] = errorsMessages['any.required']
+        }
+        break
+      case 'additionalProperties':
+        reasons[params.additionalProperty] = errorsMessages['object.unknown']
+        break
+      default:
+        reasons[name] = getCustomErrorMessage(schema, schemaPath, message)
+    }
+    return reasons
+  }, {})
+}
+
+const formatStack = error => get(error, 'stack', '')
+  .split('\n')
+  .filter((s, i) => i > 0)
+  .map(s => s.trim().replace(/^at /, ''))
+
+const validationContextMessages = {
+  querystring: 'Invalid input data.',
+  body: 'Invalid input data.'
+}
+
+function formatValidationErrorResponse (error, context) {
+  const { validationContext } = error
+  const schema = context.schema[validationContext]
+  return {
+    statusCode: 422,
+    error: 'Unprocessable Entity',
+    message: validationContextMessages[validationContext],
+    reasons: formatReasons(error, schema)
+  }
+}
+
+async function trail (server, options) {
+  const trailsManager = new TrailsManager(undefined, options.pool)
+
+  server.decorate('trailCore', trailsManager)
+  server.decorateReply('trailCore', trailsManager)
+
+  server.addHook('onClose', async (instance, done) => {
+    await trailsManager.close()
+    done()
+  })
+
+  const ajv = new Ajv({ allErrors: true })
```

Comment:
why is this needed?

Reply:
So that all of Ajv's configuration options are easily available. Wasn't clear at start which ones I would need to change, and left this in as may be useful later in the project to change the settings.